### ISSUE TITLE
Problem rescheduling POD with GCE PD disk attached

### ISSUE THREAD

**Issue by rvrignaud** [kubernetes/kubernetes#14642]
Hello,

I'm using a GKE (1.0.6) cluster. Today, for a yet unknown reason, a node has rebooted. This node used to have a pod with a GCE PD attached. This pod is scheduled by a RC with only one replica.
When the node rebooted, the pod has been rescheduled on an other node. However for some reason, the PD has not been detached from the old node.
The result was that Kubernetes tried multiple time to attach the disk to the new node. I got a lot of errors in GCE Operations Dashboard:
```
RESOURCE_IN_USE_BY_ANOTHER_RESOURCE: The disk resource 'projects/projectid/zones/europe-west1-c/disks/diskname' is already being used by 'projects/projectid/zones/europe-west1-c/instances/gke-nodename' 
```
At the end, the pod is in Waiting state with that reason:
```
Image: gcr.io/projectid/imagename:imagetag is not ready on the node
```
(which is not IMHO the right error message)

And events:
```
  Mon, 28 Sep 2015 07:09:34 +0200	Mon, 28 Sep 2015 10:50:26 +0200	124	{kubelet nodename}			failedMount	Unable to mount volumes for pod "podname": Could not attach GCE PD "diskname". Timeout waiting for mount paths to be created.
  Mon, 28 Sep 2015 07:09:34 +0200	Mon, 28 Sep 2015 10:50:26 +0200	124	{kubelet nodename}			failedSync	Error syncing pod, skipping: Could not attach GCE PD "diskname". Timeout waiting for mount paths to be created.
```

As this a not critical service, I'm happy to let it for a few days in this state, if that can help for debugging.
Is there any other thing I could provide to help understand the problem ?

**Comment by saad-ali** [comment]
The sync loop in kubelet is responsible for detecting and unmounting/detaching PDs that are no longer being referenced by any pod.

The problem here is that when the node goes down (is rebooted) this logic does not get a chance to do this, therefore the PD remains attached to the node. 

The Kubernetes scheduler reschedules the pod that was previously running on the downed node to another node. The new node realizes that it needs to attach the PD for the pending pod. This process then fails because the PD is already attached to the previous node (GCE PDs can only be attached to one machine in RW mode).

Potential solutions:
* If we detect that a GCE attach call is failing because the PD is attached to a different node, add logic to automatically detach the disk from the other node so that attachment will succeed
  * Problems with this approach: could end up with a race between machines trying to attach/detach the same PD from each other.
* Force some sort of garbage collection, when kubelet is killed that automatically detaches all disks.
  * Problems with this approach: How would this works?

**Comment by thockin** [comment]
If/When the "dead" kubelet comes back, won't it unmount the PD?  This is far from ideal, obviously, but GCE VMs don't normally "die" and not come back.  But they can if there's disk corruption or something, so we'll need to handle it.

Saad and I brainstormed a bit today - look for a design doc soon.

**Comment by chrislovecnm** [comment]
Hey all. Appreciate you guys focusing on this! We are a google cloud customer, and we have had production downtime because of this bug. We are worked no with support, and have verified this.  Let us know if we can provide any information. 

Once we have a tested patch, we would like to roll it into production.

Thanks again, don't want to have this kill a node again.

**Comment by chrislovecnm** [comment]
Oh and I guess I cannot type on my phone without typos today

**Comment by chrislovecnm** [comment]
@fgrzadkowski and other google gurus. Any update on this?

**Comment by saad-ali** [comment]
@chrislovecnm I am going to send out a design proposal for this, since it is a major change (not a trivial bug fix). It is unlikely to make it into our v1.1 release, however it'll be my top priority thereafter.

**Comment by chrislovecnm** [comment]
@saad-ali that is an awesome update.  Peeps here will be happy!!  We are adding monitoring to ensure that we catch this if we have another problem like this.  Let me know when you think you will have a time estimate ... I know I know ... my bosses are asking ...

**Comment by saad-ali** [comment]
@chrislovecnm I can't comment on an ETA, but I will say it's high pri post v1.1.

**Comment by saad-ali** [comment]
While we work on a permanent fix, for folks running into this issue, here are a couple workarounds:

1. Do nothing, wait for the downed node to come back up.
  * Once the original node is rebooted and kubelet comes up on it, it will realize that the pod/volume are no longer scheduled on it and will detach the PD. Then the other node will then be able to attach it.
  * This assumes that the original, downed node will come back up..
2. Manually detach the disk.
  * Manually detach the disk using the [gcloud API](https://cloud.google.com/compute/docs/disks/persistent-disks#detach_a_non-root_persistent_disk) (CLI or REST or WebUI) from the downed node, and this will allow the other node to attach it.

**Comment by rvrignaud** [comment]
@saad-ali im my case the node come back up and kubelet did not detach the PD even if the pod was rescheduled in an other node. I had to detach it manually.

**Comment by paralin** [comment]
Just ran into this.

**Comment by chrislovecnm** [comment]
@saad-ali your release is out.  What is the status on releasing this?

**Comment by saad-ali** [comment]
@chrislovecnm Yes v1.1 is out. I'm working on the detailed design for this at the moment. I can't comment on a release timeline.

**Comment by MaxDaten** [comment]
Is this issue related to my problem with RC's or do I have a fundamental misunderstanding of PV & PVC with RCs? Or is my problem different and a new issue?
https://github.com/MaxDaten/k8s-pvc-rc

**Comment by chrislovecnm** [comment]
@saad-ali thanks ... went through a update on the production cluster, and this HAMMERED us again.... :(

**Comment by saad-ali** [comment]
@chrislovecnm Can you send an email to me (email address in profile). I'd like to better understand why you guys are running into this so often, and if there's anything we can do in the immediate term to work around it. Thanks!

**Comment by chrislovecnm** [comment]
@saad-ali it was on an update gcloud command :) We are not planning on running the update again, any time shortly.

**Comment by housebolt** [comment]
+1, this has been a persistent problem with our internal apps and has made me very wary of moving any of our critical production to kube/GC. During cluster updates or after a node goes down, I have no guarantee that PD's will attach correctly, and that leaves us open to an unacceptable level of potential downtime

**Comment by johanhaleby** [comment]
We're also experiencing this problem: 

```bash
 FirstSeen	LastSeen	Count	From					SubobjectPath	Reason		Message
  ─────────	────────	─────	────					─────────────	──────		───────
  1h		1m		78	{kubelet gke-xxxxxxx-node-ky8x}			FailedMount	Unable to mount volumes for pod "some-pod-124-noyvz_default": Could not attach GCE PD "mongodb-disk". Timeout waiting for mount paths to be created.
  1h		1m		78	{kubelet gke-yyyyyyy-117b8ac5-node-ky8x}			FailedSync	Error syncing pod, skipping: Could not attach GCE PD "mongodb-disk". Timeout waiting for mount paths to be created.
```

The disk seems to be fine though (`gcloud compute disks list`):

```
NAME                        ZONE           SIZE_GB TYPE        STATUS
mongodb-disk            europe-west1-c 1       pd-standard READY
```

**Comment by saad-ali** [comment]
We're [working on a fix](https://github.com/kubernetes/kubernetes/issues/20262). It is targeted for the next minor release v1.3.0.

**Comment by johanhaleby** [comment]
Nice to see that this is being working on. But what do we do in the meantime? We have a disk that we can't seem to mount that contains data for mongodb. I've tried deleting the pods and replication controller but whatever I try I can't seem to get the disk to be mounted. Luckily this happened in our test environment but we're running this is production as well and if it happens there it's quite bad.

**Comment by Spark2Ignite** [comment]
@saad-ali isn't the timeframe for v1.3.0 is at least 5 months away?

**Comment by alexcouper** [comment]
@johanhaleby I see you're running on GKE. When i've had this issue in the past it's because the disk is mounted on a different node in the cluster than the one that the pod is booting on
I went into the google dev console and unmounted manually from the node and then recreated the RC for the pod and all worked.

**Comment by saad-ali** [comment]
> what do we do in the meantime? 

@johanhaleby What @alexcouper recommended: detach the disk manually via gcloud API, CLI, or web UI that should unclog the pipes.

> @saad-ali isn't the timeframe for v1.3.0 is at least 5 months away?

The next minor release is likely 2-5 months out. That said, if we have something before then that unblocks folks, we'll patch it back into an intermediate 1.2.x release, so it may be sooner than that--but no promises.

**Comment by johanhaleby** [comment]
@saad-ali @alexcouper Thanks that seems to work

**Comment by chrislovecnm** [comment]
@saad-ali ouch man ... 2-5 months? This has been open a long time.  I assume that this problem has caused a major re-write?  This is still on my oh crap tracking list with c-levels, please let us know when you have a closer ETA.  We have not hit it again, but I am waiting for the next time it nips us ... Having mysql not restart gracefully is ZERO fun.

**Comment by saad-ali** [comment]
I hear ya @chrislovecnm. Hang tight! We're working on https://github.com/kubernetes/kubernetes/issues/21931 It'll be part of 1.2.X and will greatly reduce the bite of this issue, by gracefully handling volumes on restart.  And the new controller shouldn't be too much longer either (i.e. before 1.3 release).

**Comment by wangjia184** [comment]
+1

`gcloud compute instances detach-disk NODE --disk DISK`

**Comment by bussyjd** [comment]
I ran in this problem and as my solution might not be related but well...
I was migrating my db to a new cluster located in another region and solved it moving the volume to the same region as the new k8s cluster
`gcloud compute disks move [disk] --destination-zone [zone]`
Hope that helps

**Comment by saad-ali** [comment]
@bussyjd That's a different issue. GCE requires that a PD must be in the same zone as the instance that will use it.

**Comment by discostur** [comment]
I think i ran into the same issue trying two deploy a pod with a GCE PD and a replication controller set to replicas=2.

Regarding the documentation that should be possible:

> Using a PD on a pod controlled by a ReplicationController will fail unless the PD is read-only or the replica count is 0 or 1.

So i setup a new cluster with one minion and Kubernetes 1.2.2. Then i created a PD, PersistentVolumes and my ReplicationController:

---
apiVersion: v1
kind: PersistentVolume
metadata:
  name: volume-nginx-data-disk 
spec:
  capacity:
    storage: 1Gi
  accessModes:
    - ReadOnlyMany
  gcePersistentDisk:
    pdName: nginx-data-disk
    fsType: ext4

---
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: claim-nginx-data-disk 
spec:
  accessModes:
    - ReadOnlyMany
  resources:
    requests:
      storage: 1Gi

---
apiVersion: v1
kind: ReplicationController
metadata:
  name: sslproxy-rc 
  labels:
    name: sslproxy-rc
spec:
  replicas: 2 
  selector:
    name: sslproxy-rc
  template:
    metadata:
      labels:
        name: sslproxy-rc
    spec:
      containers:
      - name: nginx
        image: nginx
        ports:
        - name: nginx-ssl
          containerPort: 443
        volumeMounts:
        - name: nginx-data-disk
          mountPath: /etc/nginx/conf.d
      volumes:
        - name: nginx-data-disk
          persistentVolumeClaim:
            claimName: claim-nginx-data-disk
            readOnly: true
---

So everything should be mounted in "readOnly" mode. When i try to deploy my RC, only the first pod is created, the second is always stuck with status "Pending".

**kubectl describe pod xyz:**

Unable to mount volumes for pod "sslproxy-rc-zxe52_default(a532eb8f-07f2-11e6-8707-42010af00108)": Could not attach GCE PD "nginx-data-disk". Timeout waiting for mount paths to be created.

{kubelet gke-sslproxy-cluster1-default-pool-4091b8f3-wys3}			Warning		FailedSync	Error syncing pod, skipping: Could not attach GCE PD "nginx-data-disk". Timeout waiting for mount paths to be created.

**kubelet.log:**

GCE operation failed: googleapi: Error 400: The disk resource 'nginx-data-disk' is already being used by 'gke-sslproxy-cluster1-default-pool-4091b8f3-wys3'
gce_util.go:187] Error attaching PD "nginx-data-disk": googleapi: Error 400: The disk resource 'nginx-data-disk' is already being used by 'gke-sslproxy-cluster1-default-pool-4091b8f3-wys3'


I think that is because of  (https://github.com/kubernetes/kubernetes/issues/21931)

> When two or more pods specify the same volume (allowed for some plugins in certain cases), the second pod will fail to start because because the volume attach call for the 2nd pod will continuously fail since the volume is already attached (by the first pod).

Kilian

**Comment by saad-ali** [comment]
@discostur That's a known symptom of the same issue.

**Comment by ScyDev** [comment]
We've also experienced this multiple times. In our case it wasn't a new node that had a problem because the disk was not properly detached from another node. 
This node existed for weeks. It was like the node was complaining that it couldn't attach the disk because it was already attached to itself.

Stackdriver log for node `gke-test-cluster-default-pool-a7afbbdb-i0sb`
```
11:28:01.000
GCE operation failed: googleapi: Error 400: The disk resource 'test-mongo-disk-test' is already being used by 'gke-test-cluster-default-pool-a7afbbdb-i0sb'
{
  metadata: {…}   
  insertId: "2016-05-20|02:28:03.111393-07|10.194.248.228|101325354"   
  log: "kubelet"   
  structPayload: {…}   
}
```

Manually detaching the disk in question helps: https://cloud.google.com/sdk/gcloud/reference/compute/instances/detach-disk

After that the node can reattach the disk properly and the pod can start again.

**Comment by saad-ali** [comment]
@ScyDev That's https://github.com/kubernetes/kubernetes/issues/19953 Should be fixed at the same time as this.

**Comment by chrislovecnm** [comment]
@saad-ali is this getting fixed in 1.3?? What is going on with this never ending issue 😄

**Comment by saad-ali** [comment]
It is indeed my friend https://github.com/kubernetes/kubernetes/issues/25457
Hang tight =)

**Comment by nikunjchapadia** [comment]
+1
